src/service/url-expander/expander.js

```diff
@@ -0,0 +1,264 @@
+    if (!url.length) {
+      return Promise.resolve(url);
+    }
+    const expr = this.variableSource_
+        .getExpr(opt_bindings, /*opt_ignoreArgs */ true);
+
+    const matches = this.findMatches_(url, expr);
+    // if no keywords move on
+    if (!matches.length) {
+      return Promise.resolve(url);
+    }
+    const mergedPositions = this.eliminateOverlaps_(matches, url,
+        opt_whiteList);
+    return this.parseUrlRecursively_(url, mergedPositions, opt_bindings);
+  }
+
+  /**
+   * Structures the regex matching into the desired format
+   * @param {string} url url to be substituted
+   * @param {RegExp} expression regex containing all keywords
+   * @return {!Object<string, string|number>} array of objects representing
+   *  matching keywords
+   */
+  findMatches_(url, expression) {
+    const matches = [];
+    url.replace(expression, (match, name, startPosition) => {
+      const length = match.length;
+      const stopPosition = length + startPosition - 1;
+      const info = {
+        start: startPosition,
+        stop: stopPosition,
+        name,
+        length,
+      };
+      matches.push(info);
+    });
+    return matches;
+  }
+
+  /**
+   * Takes array of all matching variable keyword locations and returns only those that we
+   * want to evaluate. In the case of overlapping keywords it choose the longer. It also
+   * excludes protected keywords.
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
+    if (!matches.length) { return null; }
+    // longest keywords have priority over shorter
+    // if same length alphabetical precidence
+    matches.sort((a, b) => {
+      if (a.length === b.length) {
+        return a.name > b.name;
+      }
+      return b.length - a.length;
+    });
+
+    const keywordMarker = new Array(url.length).fill(false);
+
+    return matches.reduce((results, match) => {
+      if (opt_whiteList && !opt_whiteList[match.name]) {
+        // Do not perform substitution and just return back the original
+        // match, so that the string doesn't change.
+        return results;
+      }
+      if (this.fillStorage_(match, keywordMarker)) {
+        // if this match does not overlap with a longer match
+        results.push(match);
+      }
+      return results;
+    }, []).sort((a, b) => a.start - b.start);
+  }
+
+  /**
+   * Given a keyword marking data structure and a match object, it will keep track of
+   * indexes that are already being tracked.
+   * @param {!Object<string, *>} match contains match data
+   * @return {boolean} indicates whether a collision was found
+   */
+  fillStorage_(match, keywordMarker) {
+    for (let i = match.start; i <= match.stop; i++) {
+      if (keywordMarker[i]) {
+        // toggle off what we thought was to be filled
```

Comment:
ha, now I think the other algorithm is cheaper:

```js
for (let i = 0; i < usedRanges.length; i++) {
  if (hasOverlap(match, usedRanges[i])) {return false;}
});
return true;

```

Reply:
the algorithm for merging the overlaps? or the other expanding algorithm?

Reply:
never mind, just figured out what you are saying

Reply:
part of deleted sorting code